### PR TITLE
upgrade to phpunit 10

### LINKED ISSUE #70: [RFC]: PHPUnit 10 support
## RFC

| Q                   | A      |
| ------------------- | ------ |
| Proposed Version(s) | 5.0.0  |
| BC Break?           | Yes |

### Goal

Upgrade the component to PHPUnit 10

### Considerations

PHPUnit dropped support for PHP 8.0 so we will have to drop it as well.

### PR DESCRIPTION
**Pull request by rohm1**
PR for #70

### CODE REVIEW

File: phpunit.xml.dist [hunk 1]
```diff
@@ -1,10 +1,11 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <phpunit
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-    xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
+    xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.0/phpunit.xsd"
```
Comment (froschdesign) [inline phpunit.xml.dist]: The version number is already managed via the entry in the Composer configuration file, so it is not necessary to add it here.
And the XML schema definition is included in the package itself: https://github.com/sebastianbergmann/phpunit/blob/844d06625f2e8bd40f73e75e220ab1601a0478f8/phpunit.xsd#L2

Comment (rohm1) [reply phpunit.xml.dist]: :+1: I upgraded the XML with `--migrate-configuration` and it generated this, I'll rollback the change

**Comment by rohm1** [comment]
anyone an idea why the tests are failing? It looks like the wrong `symfony/dom-crawler` version is used. In my branch it's using 6.0.17 (just like in 4.8.x branch) (https://github.com/rohm1/laminas-test/blob/phpunit-10/composer.lock#L2776) but the CI job downgrades to 5.4 (https://github.com/laminas/laminas-test/actions/runs/4084028062/jobs/7040249864#step:3:114)

**Comment by gsteel** [comment]
CI runs a job for each PHP version with `composer update --prefer-lowest` to test on lowest deps. You'll need to update the composer constraints for the Symfony deps.

File: composer.json [hunk 2]
```diff
@@ -18,7 +18,7 @@
     "config": {
         "sort-packages": true,
         "platform": {
-            "php": "8.0.99"
+            "php": "8.1.0"
```
Comment (gsteel) [inline composer.json]: This should be 8.1.99

**Comment by rohm1** [comment]
@gsteel I see. Thanks!

File: test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php [hunk 3]
```diff
@@ -240,9 +241,17 @@ public function testAssertXpathQueryWithBadXpathUsage(): void
...
 
-        $this->expectError();
+        // symfony/dom-crawler 5.4: InvalidArgumentException
+        // symfony/dom-crawler 6.0: ExpectationFailedException
 
-        $this->assertXpathQuery('form#myform');
+        try {
+            $this->assertXpathQuery('form#myform');
+        } catch (Throwable) {
+            $this->assertTrue(true);
```
Comment (gsteel) [inline test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php]: Better to assert the expected exception type here. Something like: 

```suggestion
        } catch (Throwable $e) {
            $this->assertTrue($e instanceof InvalidArgumentException || $e instanceof ExpectationFailedException);
```

File: test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php [hunk 4]
```diff
@@ -240,9 +241,17 @@ public function testAssertXpathQueryWithBadXpathUsage(): void
...
+            return;
+        }
+
+        $this->fail('should not have been called');
```
Comment (gsteel) [inline test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php]: A more descriptive failure message would be helpful. An exception is expected (Assuming DOM Crawler is responsible for throwing this).

```suggestion
        $this->fail('An exception should have been thrown by Symfony DOM Crawler');
```

File: phpunit.xml.dist [hunk 5]
```diff
@@ -3,8 +3,9 @@
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
     bootstrap="./test/bootstrap.php"
-    colors="true">
-    <coverage processUncoveredFiles="true">
+    colors="true"
+    cacheDirectory=".phpunit.cache">
```
Comment (gsteel) [inline phpunit.xml.dist]: Currently `.phpunit.result.cache` is listed in `.gitignore` - if this is a change with PHPUnit 10, can you update .gitignore instead, or make both things match?

**Review by gsteel — CHANGES_REQUESTED** [review]
Nice! Couple of minor tweaks from me

File: test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php (hunk 3 above)
Comment (rohm1) [reply test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php]: I've changed it to two catch blocks with message check

**Comment by rohm1** [comment]
> Nice! Couple of minor tweaks from me

thanks for the review!

**Review by gsteel — APPROVED** [review]
LGTM! @ocramius, can this go straight into a minor release if you approve?

File: phpunit.xml.dist [hunk 6]
```diff
@@ -3,8 +3,9 @@
     xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
     xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
     bootstrap="./test/bootstrap.php"
-    colors="true">
-    <coverage processUncoveredFiles="true">
+    colors="true"
+    cacheDirectory=".phpunit.result.cache">
```
Comment (Ocramius) [inline phpunit.xml.dist]: Can be omitted

File: src/PHPUnit/Controller/AbstractHttpControllerTestCase.php [hunk 7]
```diff
@@ -426,7 +426,7 @@ private function query($path, $useXpath = false)
             }
         }
 
-        return $useXpath ? $document->filterXPath($path) : $document->filter($path);
+        return $useXpath ? @$document->filterXPath($path) : $document->filter($path);
```
Comment (Ocramius) [inline src/PHPUnit/Controller/AbstractHttpControllerTestCase.php]: This `@` should not be here.

File: test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php [hunk 8]
```diff
@@ -240,9 +241,17 @@ public function testAssertXpathQueryWithBadXpathUsage(): void
     {
         $this->dispatch('/tests');
 
-        $this->expectError();
+        // symfony/dom-crawler 5.4: InvalidArgumentException
+        // symfony/dom-crawler 6.0: can process the path, but path invalid so nothing is found:
```
Comment (Ocramius) [inline test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php]: Let's please drop 5.4 :)

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php (hunk 8 above)
Comment (rohm1) [reply test/PHPUnit/Controller/AbstractHttpControllerTestCaseTest.php]: ok :)

**Review by Ocramius — APPROVED** [review]
Thanks @rohm1!

**Comment by rohm1** [comment]
Thanks guys!

**Pull request merged**